### DISCUSSION TITLE
Output of postUpgradeTasks

### DISCUSSION THREAD

**Discussion by nsheaps** [renovatebot/renovate#17418] [Request Help]
I've noticed when a post upgrade task fails it posts in the upgrade PR some of the failure output. I have some functionality that runs through an updater script to update other chunks when a specific package upgrades, and that tool outputs info that I'd like included on the PR (since it includes things like testing instructions or what was changed by the upgrader). I see the output in the logs, but I don't see a way to configure how output is handled for post upgrade tasks. Is there a renovate-native way to do this? if not I was considering using the `gh` utility to post a comment back to the PR, however that requires the PR be open first (currently we have it configured to not open PRs until checks have run)

Using renovate on prem if that's helpful. We have some post upgrade scripts baked into the image for ease of use with some specific tasks.

**Comment by rarkins** [comment]
This behaviour is intentional. For example today you don't see CI systems posting the entire CI log to a PR where tests fail. Stdout can be too long to fit into a PR comment. 

Both stdout and stderr should be included in logs. Only stderr is posted to the PR itself. Is your problem that repo users don't have access to the logs?

**Reply by nsheaps** [reply]
Generally the logs can be quite noisy and hard to follow when enabled on a lot of repos and users are unfamiliar with it, only a few of us actually look at the logs.

As for the stdout error output from a failed task, that's easy enough to change, but misses the rest of the output that guides the user as to what it found/where/why it's doing things. Realistically, you were correct to close my proposed PR, I don't even want _all_ the output from the commands I'm running, just one of them

Realistically, we are looking to get some output (not all) from the commands into a PR to provide post upgrade instructions. We're considering using `gh pr comment` since it's not immediately supported natively, but I think there's a risk there that the postUpgradeCommands run before the PR is created. I'm having a bit of trouble following the logic [here](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/index.ts#L105) with relation to when the PR is actually created.

Rather than trying to potentially hack it into place into the commands itself, what would you think about adding a new feature (I can contribute code + docs) where commands can either be a string as they are now, or an object like the following:

```
....commands: [
  "command a",
  {cmd:"command b", output:"pr", stderr:true, stdout:true, includeStdoutOnError:true}
]
```

`cmd` would be the command string as it exists now

`output` would be `none` or `pr`

`stderr` and `stdout` would control what's posted, would post in the initial PR body, concatenating by all the commands that utilize it

`includeStdoutOnError` would allow the failure stdout to get posted to the comment that's posted to the PR.
